File: hackerrank/problem solving/algorithms/2 - implementation/28 - jumping on the clouds - revisited/jumping_on_the_clouds_revisited_procedural.py
# Source: https://www.hackerrank.com/challenges/jumping-on-the-clouds-revisited/problem?isFullScreen=true

def main():

    array = readAnArray()
    _SIZE_ARRAY = array[0]
    JUMP_LENGTH = array[1]
    array = readAnArray()

    ENERGY = jumpingOnClouds(array, JUMP_LENGTH)
    print(ENERGY)


def readAnArray():

    ARRAY = list(map(int, input().split()))
    return ARRAY


def jumpingOnClouds(array, jumpLength):

    energy = 100

    cloudIndex = 0
    while True:
        energy -= spentEnergyAccordingToTypeOfCloud(array[cloudIndex])
        cloudIndex = generateNewCloudIndex(cloudIndex, len(array), jumpLength)
        if isCloudIndexBackToFirstCloud(cloudIndex):
            break

    return energy


def spentEnergyAccordingToTypeOfCloud(cloudType):

    return 1 if cloudType == 0 else 3


def generateNewCloudIndex(cloudIndex, sizeArray, jumpLength):

    return (cloudIndex + jumpLength) % sizeArray


def isCloudIndexBackToFirstCloud(cloudIndex):

    return cloudIndex == 0


if __name__ == "__main__":
    main()

File: hackerrank/problem solving/algorithms/2 - implementation/28 - jumping on the clouds - revisited/jumping_on_the_clouds_revisited_procedural.js
// Source: https://www.hackerrank.com/challenges/jumping-on-the-clouds-revisited/problem?isFullScreen=true

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', function(inputStdin) {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.split('\n');
    main();
});

function readLine() {
    return inputString[currentLine++];
}

function main() {
    let array = readAnArray();
    const _SIZE_ARRAY = array[0];
    const JUMP_LENGTH = array[1];
    array = readAnArray();

    const ENERGY = jumpingOnClouds(array, JUMP_LENGTH);
    console.log(ENERGY);
}

    function readAnArray() {
        const ARRAY = readLine().split(" ").map(Number);
        return ARRAY;
    }

    function jumpingOnClouds(array, jumpLength) {
        let energy = 100;

        for (let cloudIndex = 0; true; ) {
            energy -= spentEnergyAccordingToTypeOfCloud(array[cloudIndex]);
            cloudIndex = generateNewCloudIndex(cloudIndex, array.length, jumpLength);
            if (isCloudIndexBackToFirstCloud(cloudIndex)) {
                break;
            }
        }

        return energy;
    }

        function spentEnergyAccordingToTypeOfCloud(cloudType) {
            return cloudType == 0 ? 1 : 3;
        }

        function generateNewCloudIndex(cloudIndex, sizeArray, jumpLength) {
            return (cloudIndex + jumpLength) % sizeArray;
        }

        function isCloudIndexBackToFirstCloud(cloudIndex) {
            return cloudIndex == 0;
        }
